File: tests/test_tags.cpp
#include <gtest/gtest.h>
#include "tags.h"

TEST(SanitizeTests, ValidTagSequence)
{
    TagCharSanitizer tag_char_sanitizer;

    std::vector<char32_t> text = {
        0x1F3F4,
        0xE0000 + 'g', 0xE0000 + 'b', 0xE0000 + 'e', 0xE0000 + 'n', 0xE0000 + 'g',
        0xE007F};

    tag_char_sanitizer.sanitize(text);

    std::vector<char32_t> expected = {
        0x1F3F4,
        0xE0000 + 'g', 0xE0000 + 'b', 0xE0000 + 'e', 0xE0000 + 'n', 0xE0000 + 'g',
        0xE007F};

    EXPECT_EQ(text, expected);
}

TEST(SanitizeTests, TagOutsideBase)
{
    TagCharSanitizer tag_char_sanitizer;
    std::vector<char32_t> text2 = {
        0xE0000 + 'g', 0xE0000 + 'b', 0xE0000 + 'e', 0xE0000 + 'n', 0xE0000 + 'g'};
    tag_char_sanitizer.sanitize(text2);

    // Invalid tag sequence should be removed
    std::vector<char32_t> expected = {};
    EXPECT_EQ(text2, expected);
}

TEST(SanitizeTests, InvalidTagSequence)
{
    TagCharSanitizer tag_char_sanitizer;
    std::vector<char32_t> text2 = {
        0x1F3F4,
        0xE0000 + 'a', 0xE0000 + 'b', 0xE0000 + 'c',
        0xE007F};
    tag_char_sanitizer.sanitize(text2);

    // Invalid tag sequence should be removed
    std::vector<char32_t> expected = {};
    EXPECT_EQ(text2, expected);
}

TEST(SanitizeTests, NoTags)
{
    TagCharSanitizer tag_char_sanitizer;
    std::vector<char32_t> text3 = {'h', 'e', 'l', 'l', 'o', ' ', 'w', 'o', 'r', 'l', 'd'};
    tag_char_sanitizer.sanitize(text3);

    // Should remain unchanged
    std::vector<char32_t> expected = {'h', 'e', 'l', 'l', 'o', ' ', 'w', 'o', 'r', 'l', 'd'};
    EXPECT_EQ(text3, expected);
}

TEST(SanitizeTests, TooLongTagSequence)
{
    TagCharSanitizer tag_char_sanitizer;
    std::vector<char32_t> text4 = {0x1F3F4};
    for (int i = 0; i < 33; i++) // 33 > MAX_TAG_LENGTH
        text4.push_back(0xE0000 + 'a');
    text4.push_back(0xE007F);

    tag_char_sanitizer.sanitize(text4);

    // Excessively long tag sequence should be removed
    std::vector<char32_t> expected = {};
    EXPECT_EQ(text4, expected);
}

File: python_tests/test_tags.py
"""
Python tests for TagCharSanitizer based on C++ test cases.
"""

import pytest

try:
    import textual_adversarial_defense
except ImportError:
    pytest.skip(
        "textual_adversarial_defense package not available", allow_module_level=True
    )


class TestTagSanitizer:
    """Test cases for tag character sanitization functionality."""

    def test_valid_tag_sequence(self):
        """Test that valid tag sequences are preserved."""
        pipeline = textual_adversarial_defense._pipeline.Pipeline()
        pipeline.add_tag_sanitizer()

        # Valid emoji flag tag sequence for England (gbeng)
        text = (
            chr(0x1F3F4)  # Black flag emoji
            + chr(0xE0000 + ord("g"))
            + chr(0xE0000 + ord("b"))
            + chr(0xE0000 + ord("e"))
            + chr(0xE0000 + ord("n"))
            + chr(0xE0000 + ord("g"))
            + chr(0xE007F)
        )  # Cancel tag

        result = pipeline.sanitize(text)
        expected = text  # Should remain unchanged

        assert result == expected, f"Expected valid tag sequence to be preserved"

    def test_tag_outside_base(self):
        """Test that tag characters without base are removed."""
        pipeline = textual_adversarial_defense._pipeline.Pipeline()
        pipeline.add_tag_sanitizer()

        # Tags without base character
        text = (
            chr(0xE0000 + ord("g"))
            + chr(0xE0000 + ord("b"))
            + chr(0xE0000 + ord("e"))
            + chr(0xE0000 + ord("n"))
            + chr(0xE0000 + ord("g"))
        )

        result = pipeline.sanitize(text)
        expected = ""  # Should be removed

        assert result == expected, f"Expected empty string, got '{result}'"

    def test_invalid_tag_sequence(self):
        """Test that invalid tag sequences are removed."""
        pipeline = textual_adversarial_defense._pipeline.Pipeline()
        pipeline.add_tag_sanitizer()

        # Invalid tag sequence (abc is not a valid sequence)
        text = (
            chr(0x1F3F4)  # Black flag emoji
            + chr(0xE0000 + ord("a"))
            + chr(0xE0000 + ord("b"))
            + chr(0xE0000 + ord("c"))
            + chr(0xE007F)
        )  # Cancel tag

        result = pipeline.sanitize(text)
        expected = ""  # Invalid sequence should be removed

        assert result == expected, f"Expected empty string, got '{result}'"

    def test_no_tags(self):
        """Test that text without tags remains unchanged."""
        pipeline = textual_adversarial_defense._pipeline.Pipeline()
        pipeline.add_tag_sanitizer()

        text = "hello world"
        result = pipeline.sanitize(text)
        expected = "hello world"

        assert result == expected, f"Expected '{expected}', got '{result}'"

    def test_too_long_tag_sequence(self):
        """Test that excessively long tag sequences are removed."""
        pipeline = textual_adversarial_defense._pipeline.Pipeline()
        pipeline.add_tag_sanitizer()

        # Create a sequence longer than MAX_TAG_LENGTH (33 tags)
        text = chr(0x1F3F4)  # Base character
        for i in range(33):  # 33 > MAX_TAG_LENGTH
            text += chr(0xE0000 + ord("a"))
        text += chr(0xE007F)  # Cancel tag

        result = pipeline.sanitize(text)
        expected = ""  # Should be removed

        assert result == expected, f"Expected empty string, got '{result}'"
